app/plugin_api.go

```diff
@@ -369,6 +369,39 @@ func (api *PluginAPI) GetUserCustomStatus(userID string) (*model.CustomStatus, *
 	return api.app.GetCustomStatus(userID)
 }
 
+func (api *PluginAPI) RestoreToPreviousCustomStatus(userID string) *model.AppError {
+	cs, appErr := api.app.GetCustomStatus(userID)
+	if appErr != nil {
+		return appErr
+	}
+
+	// If user changed the custom status explicitly, don't restore to previous status
+	if cs == nil || !cs.SetByProduct {
+		return nil
+	}
+
+	rcs, err := api.app.GetRecentCustomStatuses(userID)
+	if err != nil {
+		return model.NewAppError("RestoreToPreviousCustomStatus", "plugin.api.restore_to_previous_custom_status", nil, "failed to get recent statuses to reset", http.StatusInternalServerError).Wrap(err)
+	}
+
+	if len(rcs) == 0 {
+		err := api.app.RemoveCustomStatus(api.ctx, userID)
+		if err != nil {
+			return model.NewAppError("RestoreToPreviousCustomStatus", "plugin.api.restore_to_previous_custom_status", nil, "failed to remove custom status", http.StatusInternalServerError).Wrap(err)
+		}
+
+		return nil
+	}
+
+	cserr := api.app.SetCustomStatus(api.ctx, userID, rcs[0])
```

Comment:
Let's say I have recent custom statuses, but my last choice as a user was to remove my custom status. With the logic here, will this function set my status back to my last recent one, rather than setting it back to my proper previous state of "no custom status"?

Reply:
Yes, https://github.com/mattermost/mattermost-server/blob/6a5a50c67490d0173b0f005d4683222a289acdb9/app/plugin_api.go#L389

Reply:
I read this code you're linking to, but I figured in my scenario described above, `len(rcs)` would not be zero. Under what conditions does `GetRecentCustomStatuses` return an empty slice?

If a user clears their status, wouldn't their "recent custom statuses" preference still contain values, thus making it so `len` is greater than zero?

Reply:
Do you mind writing a unit test for this block? The steps I'm thinking:

- User sets custom status
- User clears custom status
- Plugin sets custom status
- Call restore status
- Assert that user has no status, rather than the initial one they had on

Reply:
@mickmister yes the more i think about it , you are right even if user clears custom status the recent ones are still there. No restoring but actually take up the first of them and set it up. What do you think we can do here?

Reply:
gentle ping on this one @mickmister

Reply:
@M-ZubairAhmed I'm thinking the recent statuses field should resemble "the most recent statuses set by the user", and shouldn't be something where we remove from that list. I don't think it should be used as a state variable for logic like this snippet we're commenting on.

Maybe we can add two fields to custom status:
- One that means "last status user manually chose", and clear this value when they clear their status
- One that means "last status set by a product", and clear this value when:
    - Product wants to clear it
    - and when User sets or clears their status manually

The product status field would always take precedence for what is shown in the UI, but it always gets cleared whenever the user sets or clears their custom status. What do you think @M-ZubairAhmed?